lib/PhasarLLVM/ControlFlow/LLVMBasedICFG.cpp

```diff
@@ -234,6 +250,85 @@ void LLVMBasedICFG::constructionWalker(const llvm::Function *F,
   }
 }
 
+bool LLVMBasedICFG::constructDynamicCall(const llvm::Instruction *I,
+                                         Resolver &Resolver) {
+  PAMM_GET_INSTANCE;
+  bool NewTargetsFound = false;
+  // Find vertex of calling function.
+  vertex_t ThisFunctionVertexDescriptor;
+  auto FvmItr = FunctionVertexMap.find(I->getParent()->getParent());
+  if (FvmItr != FunctionVertexMap.end()) {
+    ThisFunctionVertexDescriptor = FvmItr->second;
+  } else {
+    LOG_IF_ENABLE(
+        BOOST_LOG_SEV(lg::get(), ERROR)
+        << "constructDynamicCall: Did not find vertex of calling function "
+        << I->getParent()->getParent()->getName().str() << " at callsite "
+        << llvmIRToString(I));
+    std::terminate();
+  }
+
+  if (llvm::isa<llvm::CallInst>(I) || llvm::isa<llvm::InvokeInst>(I)) {
+    Resolver.preCall(I);
+    llvm::ImmutableCallSite CS(I);
+    set<const llvm::Function *> PossibleTargets;
+    // the function call must be resolved dynamically
+    LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG)
+                  << "Looking into dynamic call-site: ");
+    LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG) << "  " << llvmIRToString(I));
+    // call the resolve routine
+    if (LLVMBasedICFG::isVirtualFunctionCall(CS.getInstruction())) {
+      PossibleTargets = Resolver.resolveVirtualCall(CS);
+    } else {
+      PossibleTargets = Resolver.resolveFunctionPointer(CS);
+    }
+    if (IndirectCalls.count(I) == 0 ||
+        IndirectCalls[I] < PossibleTargets.size()) {
+      LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG)
+                    << "Found " << PossibleTargets.size() - IndirectCalls[I]
+                    << " new possible target(s)");
+      IndirectCalls[I] = PossibleTargets.size();
+      NewTargetsFound = true;
+    }
+    if (!NewTargetsFound) {
+      return NewTargetsFound;
+    }
+    // Throw out already found targets
+    for (const auto &OE : boost::make_iterator_range(
+             boost::out_edges(ThisFunctionVertexDescriptor, CallGraph))) {
+      if (CallGraph[OE].CS == I) {
+        PossibleTargets.erase(CallGraph[boost::target(OE, CallGraph)].F);
+      }
+    }
+    Resolver.handlePossibleTargets(CS, PossibleTargets);
+    // Insert possible target inside the graph and add the link with
+    // the current function
+    for (const auto &PossibleTarget : PossibleTargets) {
+      vertex_t TargetVertex;
+      auto TargetFvmItr = FunctionVertexMap.find(PossibleTarget);
+      if (TargetFvmItr != FunctionVertexMap.end()) {
+        TargetVertex = TargetFvmItr->second;
+      } else {
+        TargetVertex =
+            boost::add_vertex(VertexProperties(PossibleTarget), CallGraph);
+        FunctionVertexMap[PossibleTarget] = TargetVertex;
+      }
+      boost::add_edge(ThisFunctionVertexDescriptor, TargetVertex,
+                      EdgeProperties(I), CallGraph);
+    }
+
+    // continue resolving
+    for (const auto *PossibleTarget : PossibleTargets) {
+      FunctionWL.push(PossibleTarget);
+    }
```

Comment:
Can you skip adding the function to the worklist if it has already been processed?
```suggestion
    // continue resolving
    for (const auto *PossibleTarget : PossibleTargets) {
      if (VisitedFunctions.count(PossibleTarget) == 0) {
          FunctionWL.push(PossibleTarget);
      }
    }
```

Reply:
Nope. The points-to information could have been updated, leading to new call edges which were not seen before.

Reply:
That makes sense.  This turned out to not be a problem for the program I was analyzing.